File: packages/contracts/headless/src/popover.rs
//! Popover machine. Mirror of core `popover.ts`.

#[path = "generated/machines/popover.rs"]
mod popover_interface;
pub use popover_interface::*;

fn open(context: PopoverContext) -> (PopoverState, Vec<PopoverEffect>) {
    (
        PopoverState::Open,
        vec![
            PopoverEffect::EmitOpenChange { open: true },
            PopoverEffect::FocusOnOpen {
                strategy: context.initial_focus,
            },
        ],
    )
}

fn close() -> (PopoverState, Vec<PopoverEffect>) {
    (
        PopoverState::Closed,
        vec![
            PopoverEffect::EmitOpenChange { open: false },
            PopoverEffect::RestoreTriggerFocus,
        ],
    )
}

pub fn popover_transition(
    state: PopoverState,
    context: PopoverContext,
    event: PopoverEvent,
) -> (PopoverState, Vec<PopoverEffect>) {
    if context.disabled {
        return (state, vec![]);
    }

    match (state, event) {
        (PopoverState::Closed, PopoverEvent::Toggle | PopoverEvent::Open) => open(context),
        (PopoverState::Open, PopoverEvent::Toggle | PopoverEvent::Close | PopoverEvent::Escape) => {
            close()
        }
        (PopoverState::Open, PopoverEvent::OutsideInteract)
            if context.dismiss_on_outside_interact =>
        {
            close()
        }
        _ => (state, vec![]),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn context(initial_focus: PopoverInitialFocus) -> PopoverContext {
        PopoverContext {
            disabled: false,
            dismiss_on_outside_interact: true,
            initial_focus,
        }
    }

    #[test]
    fn toggle_and_open_from_closed_emit_open_and_focus_effects() {
        for event in [PopoverEvent::Toggle, PopoverEvent::Open] {
            let (state, effects) = popover_transition(
                PopoverState::Closed,
                context(PopoverInitialFocus::FirstFocusable),
                event,
            );
            assert_eq!(state, PopoverState::Open);
            assert_eq!(
                effects,
                vec![
                    PopoverEffect::EmitOpenChange { open: true },
                    PopoverEffect::FocusOnOpen {
                        strategy: PopoverInitialFocus::FirstFocusable,
                    },
                ]
            );
        }
    }

    #[test]
    fn every_close_axis_restores_trigger_focus() {
        // Contract: focus returns to the trigger on every close path — toggle,
        // explicit close, Escape, and outside dismissal.
        for event in [
            PopoverEvent::Toggle,
            PopoverEvent::Close,
            PopoverEvent::Escape,
            PopoverEvent::OutsideInteract,
        ] {
            let (state, effects) = popover_transition(
                PopoverState::Open,
                context(PopoverInitialFocus::Content),
                event,
            );
            assert_eq!(state, PopoverState::Closed);
            assert_eq!(
                effects,
                vec![
                    PopoverEffect::EmitOpenChange { open: false },
                    PopoverEffect::RestoreTriggerFocus,
                ]
            );
        }
    }

    #[test]
    fn open_events_are_inert_while_closed_but_toggle_closes_an_open_surface() {
        let (state, effects) = popover_transition(
            PopoverState::Closed,
            context(PopoverInitialFocus::None),
            PopoverEvent::Close,
        );
        assert_eq!(state, PopoverState::Closed);
        assert!(effects.is_empty());

        let (state, effects) = popover_transition(
            PopoverState::Open,
            context(PopoverInitialFocus::None),
            PopoverEvent::Open,
        );
        assert_eq!(state, PopoverState::Open);
        assert!(effects.is_empty());
    }

    #[test]
    fn outside_interact_obeys_dismiss_on_outside_interact() {
        let guarded = PopoverContext {
            dismiss_on_outside_interact: false,
            ..context(PopoverInitialFocus::FirstFocusable)
        };
        let (state, effects) =
            popover_transition(PopoverState::Open, guarded, PopoverEvent::OutsideInteract);
        assert_eq!(state, PopoverState::Open);
        assert!(effects.is_empty());

        let (state, effects) = popover_transition(
            PopoverState::Open,
            context(PopoverInitialFocus::FirstFocusable),
            PopoverEvent::OutsideInteract,
        );
        assert_eq!(state, PopoverState::Closed);
        assert_eq!(effects.len(), 2);
    }

    #[test]
    fn disabled_blocks_every_open_direction_and_keeps_state() {
        // Contract: disabled blocks `setOpen` in all directions.
        let disabled = PopoverContext {
            disabled: true,
            dismiss_on_outside_interact: true,
            initial_focus: PopoverInitialFocus::FirstFocusable,
        };
        for event in [
            PopoverEvent::Toggle,
            PopoverEvent::Open,
            PopoverEvent::Close,
            PopoverEvent::Escape,
            PopoverEvent::OutsideInteract,
        ] {
            for state in [PopoverState::Closed, PopoverState::Open] {
                let (next, effects) = popover_transition(state, disabled, event);
                assert_eq!(next, state, "disabled must never change state for {event:?}");
                assert!(effects.is_empty());
            }
        }
    }

    #[test]
    fn focus_effect_carries_the_declared_strategy() {
        for strategy in [
            PopoverInitialFocus::FirstFocusable,
            PopoverInitialFocus::Content,
            PopoverInitialFocus::None,
        ] {
            let (_, effects) = popover_transition(
                PopoverState::Closed,
                context(strategy),
                PopoverEvent::Open,
            );
            assert_eq!(
                effects,
                vec![
                    PopoverEffect::EmitOpenChange { open: true },
                    PopoverEffect::FocusOnOpen { strategy },
                ]
            );
        }
    }
}

File: packages/core/src/popover.ts
/**
 * Popover behavior machine.
 * Contract: docs/contracts/components/popover.md, "Behavior Machine".
 *
 * States: closed | open. Focus restoration runs on every close path
 * (explicit, escape, outside). Dismissal wiring is the adapter's job via
 * registerDismissLayer while open; positioning stays CSS-anchored (documented
 * delta until the shared anchor-positioning service lands).
 */

import type { PartAttrs, TransitionResult } from "./machine";
import type {
  PopoverState,
  PopoverInitialFocus,
  PopoverContext,
  PopoverEvent,
  PopoverEffect,
} from "./generated/machines/popover";

export type {
  PopoverState,
  PopoverInitialFocus,
  PopoverContext,
  PopoverEvent,
  PopoverEffect,
};

export type PopoverResult = TransitionResult<PopoverState, PopoverContext, PopoverEffect>;

function open(context: PopoverContext): PopoverResult {
  return {
    state: "open",
    context,
    effects: [
      { type: "emitOpenChange", open: true },
      { type: "focusOnOpen", strategy: context.initialFocus },
    ],
  };
}

function close(context: PopoverContext): PopoverResult {
  return {
    state: "closed",
    context,
    effects: [
      { type: "emitOpenChange", open: false },
      { type: "restoreTriggerFocus" },
    ],
  };
}

export function popoverTransition(
  state: PopoverState,
  context: PopoverContext,
  event: PopoverEvent,
): PopoverResult {
  const stay: PopoverResult = { state, context, effects: [] };

  if (context.disabled) {
    return stay;
  }

  switch (event.type) {
    case "TOGGLE":
      return state === "closed" ? open(context) : close(context);
    case "OPEN":
      return state === "closed" ? open(context) : stay;
    case "CLOSE":
      return state === "open" ? close(context) : stay;
    case "ESCAPE":
      return state === "open" ? close(context) : stay;
    case "OUTSIDE_INTERACT":
      return state === "open" && context.dismissOnOutsideInteract ? close(context) : stay;
  }
}

export interface PopoverPartProps {
  surfaceId: string;
  ariaLabel?: string | null;
  block: boolean;
  triggerIsInteractive?: boolean;
  placement: string;
  surfaceWidth: "content" | "trigger";
}

/**
 * Framework-neutral trigger state the web adapters hand to the caller's real
 * interactive control in interactive mode. `expanded` follows the open state,
 * `controls` is the rendered surface id while open and `null` while closed
 * (preserving the conditional `aria-controls` contract), and `disabled`
 * mirrors the effective Popover disabled state.
 */
export type PopoverTriggerState = Readonly<{
  expanded: boolean;
  controls: string | null;
  disabled: boolean;
}>;

export interface PopoverParts {
  root: PartAttrs;
  trigger: PartAttrs;
  surface: PartAttrs;
  triggerState: PopoverTriggerState;
}

export function popoverParts(
  state: PopoverState,
  context: PopoverContext,
  props: PopoverPartProps,
): PopoverParts {
  const isOpen = state === "open";

  return {
    root: {
      "data-scope": "popover",
      "data-part": "root",
      "data-state": state,
      "data-block": props.block,
      "data-placement": props.placement,
      "data-surface-width": props.surfaceWidth,
    },
    trigger: {
      "data-part": "trigger",
      "data-state": state,
      "data-block": props.block,
      "data-disabled": context.disabled,
      role: props.triggerIsInteractive ? undefined : "button",
      tabindex: props.triggerIsInteractive ? undefined : context.disabled ? -1 : 0,
      "aria-disabled": props.triggerIsInteractive ? undefined : context.disabled ? "true" : undefined,
      "aria-expanded": props.triggerIsInteractive ? undefined : isOpen ? "true" : "false",
      "aria-controls": props.triggerIsInteractive ? undefined : isOpen ? props.surfaceId : undefined,
    },
    surface: {
      "data-part": "surface",
      "data-state": state,
      "data-placement": props.placement,
      "data-surface-width": props.surfaceWidth,
      id: props.surfaceId,
      role: "dialog",
      "aria-label": props.ariaLabel ?? undefined,
      tabindex: context.initialFocus === "content" ? 0 : -1,
    },
    triggerState: {
      expanded: isOpen,
      controls: isOpen ? props.surfaceId : null,
      disabled: context.disabled,
    },
  };
}
